src/main/java/de/tum/in/www1/artemis/service/exam/ExamService.java

```diff
@@ -663,4 +669,25 @@ public void archiveExam(Exam exam) {
 
         groupNotificationService.notifyInstructorGroupAboutExamArchiveState(exam, NotificationType.EXAM_ARCHIVE_FINISHED, exportErrors);
     }
+
+    /**
+     * Combines the template commits of all programming exercises in the exam.
+     * This is executed before the individual student exams are generated.
+     *
+     * @param exam - the exam which template commits should be combined
+     */
+    public void combineTemplateCommitsOfAllProgrammingExercisesInExam(Exam exam) {
+        exam.getExerciseGroups().forEach(group -> group.getExercises().stream().filter(exercise -> exercise instanceof ProgrammingExercise)
+                .map(exercise -> (ProgrammingExercise) exercise).forEach(exercise -> {
```

Comment:
you don't need the mapping to the ProgrammingExercise here, because you just use its `id`:

```suggestion
                .forEach(exercise -> {
```